File: comm/BIT.cpp
/**
 * Binary Indexed Tree (Fenwick Tree). Holds an array of type T.
 * T is a commutative monoid. Indices are 1 .. n.
 * Verified by AtCoder ARC043 C (http://arc043.contest.atcoder.jp/submissions/985157)
 */
template <class T, class Op> class BIT {
private:
  int n;
  std::vector<T> ary;
  Op op;
  T e;
public:
  BIT(int n, Op op = Op(), T e = T()) : op(op), e(e) {
    assert (n > 0);
    while(n != (n & (-n))) { // find the least power of 2 >= n
      n += n & (-n);
    }
    this->n = n;
    ary = std::vector<T>(n + 1);
    for(int i = 0; i <= n; i++) {
      ary[i] = e;
    }
  }
  /**
   * gets the sum in [1 .. idx]
   * @param idx
   * @return sum
   */
  T accum(int idx) {
    T sum = e;
    while(idx > 0) {
      sum = op(sum, ary[idx]);
      idx &= idx - 1;
    }
    return sum;
  }
  /**
   * performs data[idx] += val;
   */
  void add(int idx, T val) {
    assert (idx > 0);
    while(idx <= n) {
      ary[idx] = op(ary[idx], val);
      idx += idx & (-idx);
    }
  }
};

template<class T>
class TwoBIT {
  BIT<T, plus<T> > bit0, bit1;
public:
  TwoBIT(int n): bit0(n + 1), bit1(n + 1) {}
  // [l, r)
  void add_interval(int l, int r, T x) {
    bit1.add(l + 1, x);
    bit1.add(r + 1, -x);
    bit0.add(l + 1, -x * (l + 1));
    bit0.add(r + 1, x * (r + 1));
  }
  // [l, r)
  T sum_interval(int l, int r) {
    return bit1.accum(r + 1) * (r + 1) + bit0.accum(r + 1)
      - bit1.accum(l + 1) * (l + 1) - bit0.accum(l + 1);
  }
};

File: comm/datastr/BIT.rs
// Binary Indexed Tree (Fenwick Tree). Holds an array of type T.
// T is a commutative monoid. Indices are in 0..n.
// Verified by ABC285-F (https://atcoder.jp/contests/abc285/submissions/38329093)
#[derive(Clone, Debug)]
pub struct BIT<T> {
    n: usize,
    ary: Vec<T>,
    e: T,
}

impl<T: Clone + std::ops::AddAssign<T>> BIT<T> {
    pub fn new(n: usize, e: T) -> Self {
        BIT { n: n, ary: vec![e.clone(); n + 1], e: e }
    }
    // Usage: self.accum(..idx)
    fn accum(&self, idx: std::ops::RangeTo<usize>) -> T {
        let mut idx = idx.end;
        let mut sum = self.e.clone();
        while idx > 0 {
            sum += self.ary[idx].clone();
            idx &= idx - 1;
        }
        sum
    }
    // Usage: self.accum(l..r)
    #[inline(always)]
    pub fn range(&self, rng: std::ops::Range<usize>) -> T
        where T: std::ops::Sub<Output = T> {
        self.accum(..rng.end) - self.accum(..rng.start)
    }
    // performs data[idx] += val;
    // 0 <= idx, idx < n
    pub fn add<U: Clone>(&mut self, mut idx: usize, val: U)
        where T: std::ops::AddAssign<U> {
        debug_assert!(idx < self.n);
        idx += 1;
        let n = self.n;
        while idx <= n {
            self.ary[idx] += val.clone();
            idx += idx & idx.wrapping_neg();
        }
    }
    // Make sure that 0 <= idx < n.
    #[allow(unused)]
    #[inline(always)]
    pub fn get(&self, idx: usize) -> T
        where T: std::ops::Sub<Output = T> {
        debug_assert!(idx < self.n);
        self.accum(..idx + 1) - self.accum(..idx)
    }
}
/// This implementation of AddAssign is useful when you want to make a 2D BIT.
impl<T: Clone, U: Clone> std::ops::AddAssign<(usize, U)> for BIT<T>
    where T: std::ops::AddAssign<U>,
          T: std::ops::AddAssign<T> {
    fn add_assign(&mut self, (idx, val): (usize, U)) {
        self.add(idx, val);
    }
}